File: backend/app/pipeline.py
import asyncio
import logging
from datetime import datetime

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import AsyncSessionLocal
from app.models import Source, RawArticle, NewsCluster, ClusterArticle
from app.scrapers import ALL_SCRAPERS
from app.clustering import ArticleInput, cluster_articles
from app.ai_synthesis import ArticleForSynthesis, synthesize_cluster

logger = logging.getLogger(__name__)

SOURCES_METADATA = {
    "clarin": {"name": "Clarín", "url": "https://www.clarin.com", "color": "#004B87", "ideology_score": 0.3, "ideology_label": "Centro-derecha"},
    "lanacion": {"name": "La Nación", "url": "https://www.lanacion.com.ar", "color": "#1A3A5C", "ideology_score": 0.6, "ideology_label": "Centro-derecha"},
    "infobae": {"name": "Infobae", "url": "https://www.infobae.com", "color": "#E30613", "ideology_score": 0.2, "ideology_label": "Centro"},
    "pagina12": {"name": "Página 12", "url": "https://www.pagina12.com.ar", "color": "#1A1A1A", "ideology_score": -0.7, "ideology_label": "Izquierda"},
    "ambito": {"name": "Ámbito", "url": "https://www.ambito.com", "color": "#FF6B00", "ideology_score": 0.1, "ideology_label": "Centro"},
    "cronista": {"name": "El Cronista", "url": "https://www.cronista.com", "color": "#2C7BB6", "ideology_score": 0.2, "ideology_label": "Centro"},
    "perfil": {"name": "Perfil", "url": "https://www.perfil.com", "color": "#8B0000", "ideology_score": -0.1, "ideology_label": "Centro"},
    "laizquierda": {"name": "La Izquierda Diario", "url": "https://www.laizquierdadiario.com", "color": "#CC0000", "ideology_score": -1.0, "ideology_label": "Izquierda"},
}


async def ensure_sources(db: AsyncSession):
    for slug, meta in SOURCES_METADATA.items():
        result = await db.execute(select(Source).where(Source.slug == slug))
        source = result.scalar_one_or_none()
        if not source:
            source = Source(slug=slug, **meta)
            db.add(source)
    await db.commit()


async def run_scraping_pipeline():
    logger.info("Starting scraping pipeline...")
    async with AsyncSessionLocal() as db:
        await ensure_sources(db)

        result = await db.execute(select(Source).where(Source.active == True))
        sources = {s.slug: s for s in result.scalars().all()}

        all_new_articles = []
        scrape_tasks = [scraper.get_articles() for scraper in ALL_SCRAPERS]
        results = await asyncio.gather(*scrape_tasks, return_exceptions=True)

        for scraper, articles in zip(ALL_SCRAPERS, results):
            if isinstance(articles, Exception):
                logger.error(f"[{scraper.source_slug}] Scraper failed: {articles}")
                continue

            source = sources.get(scraper.source_slug)
            if not source:
                continue

            for art_data in articles:
                existing = await db.execute(
                    select(RawArticle).where(RawArticle.url == art_data.url)
                )
                if existing.scalar_one_or_none():
                    continue

                article = RawArticle(
                    source_id=source.id,
                    title=art_data.title,
                    url=art_data.url,
                    summary=art_data.summary,
                    published_at=art_data.published_at or datetime.utcnow(),
                    scraped_at=datetime.utcnow(),
                )
                db.add(article)
                all_new_articles.append((article, source.slug, source.name))

        await db.commit()
        logger.info(f"Scraped {len(all_new_articles)} new articles")

        # Refresh articles with IDs
        await db.flush()
        for art, *_ in all_new_articles:
            await db.refresh(art)

        if len(all_new_articles) < 2:
            logger.info("Not enough articles to cluster")
            return

        # Cluster
        inputs = [
            ArticleInput(
                id=art.id,
                title=art.title,
                summary=art.summary or "",
                source_slug=slug,
            )
            for art, slug, _ in all_new_articles
        ]

        clusters = cluster_articles(inputs)
        logger.info(f"Found {len(clusters)} clusters")

        slug_to_name = {slug: name for _, slug, name in all_new_articles}
        id_to_data = {art.id: (art, slug, name) for art, slug, name in all_new_articles}

        for cluster in clusters:
            arts_for_synthesis = []
            for art_id in cluster.article_ids:
                if art_id not in id_to_data:
                    continue
                art, slug, name = id_to_data[art_id]
                arts_for_synthesis.append(ArticleForSynthesis(
                    source_name=slug_to_name.get(slug, slug),
                    source_slug=slug,
                    title=art.title,
                    summary=art.summary or "",
                    url=art.url,
                ))

            synthesis = await synthesize_cluster(arts_for_synthesis)
            if not synthesis:
                continue

            news_cluster = NewsCluster(
                title=synthesis.title,
                synthesis=synthesis.synthesis,
                key_facts=synthesis.key_facts,
                category=synthesis.category,
                published_at=datetime.utcnow(),
                source_count=len(set(cluster.source_slugs)),
            )
            db.add(news_cluster)
            await db.flush()
            await db.refresh(news_cluster)

            sa_by_slug = {sa.source_slug: sa for sa in synthesis.source_analyses}

            for art_id in cluster.article_ids:
                if art_id not in id_to_data:
                    continue
                art, slug, _ = id_to_data[art_id]
                sa = sa_by_slug.get(slug)

                ca = ClusterArticle(
                    cluster_id=news_cluster.id,
                    article_id=art.id,
                    coverage_percentage=sa.coverage_percentage if sa else 0.0,
                    emphasis=sa.emphasis if sa else "",
                    omissions=sa.omissions if sa else "",
                    similarity_score=cluster.similarity_scores.get(art_id, 0.0),
                )
                db.add(ca)
                art.clustered = True

        await db.commit()
        logger.info("Pipeline complete")

File: frontend/src/lib/pipeline.ts
/* CADA LADO — Scrape → cluster → synthesize pipeline (orchestrator) */

import { and, eq, gte, desc } from 'drizzle-orm'
import { db } from './db'
import {
  sources, rawArticles, newsClusters, clusterArticles,
  type NewRawArticle,
} from './db/schema'
import { SOURCES } from './sources'
import { scrapeAllSources } from './scrapers'
import { clusterArticles as runClustering, type ArticleInput } from './clustering'
import { synthesizeCluster, type ArticleForSynthesis } from './ai-synthesis'

/* ── Seed sources (idempotent upsert) ─────────────────────────────── */
export async function ensureSources(): Promise<void> {
  for (const s of SOURCES) {
    await db.insert(sources).values({
      slug:           s.slug,
      name:           s.name,
      url:            s.url,
      rssUrl:         s.rssUrl,
      color:          s.color,
      ideologyScore:  s.ideologyScore,
      ideologyLabel:  s.ideologyLabel,
      active:         true,
    }).onConflictDoNothing({ target: sources.slug })
  }
}

/* ── Stage 1: Scrape + save raw articles ─────────────────────────── */
export async function ingestNewArticles(): Promise<number> {
  await ensureSources()

  // Build slug → id map
  const allSources = await db.select().from(sources)
  const idBySlug = new Map(allSources.map(s => [s.slug, s.id]))

  const scraped = await scrapeAllSources()
  // enrichImages skipped — fetches full HTML per article, too slow for Vercel 10s limit

  let newCount = 0
  for (const art of scraped) {
    const sourceId = idBySlug.get(art.sourceSlug)
    if (!sourceId) continue

    const row: NewRawArticle = {
      sourceId,
      title:       art.title.slice(0, 500),
      url:         art.url.slice(0, 1000),
      summary:     art.summary,
      publishedAt: art.publishedAt ?? new Date(),
      scrapedAt:   new Date(),
      imageUrl:    art.imageUrl,
      clustered:   false,
    }
    const result = await db.insert(rawArticles).values(row).onConflictDoNothing({ target: rawArticles.url }).returning({ id: rawArticles.id })
    if (result.length > 0) newCount++
  }

  console.log(`Inserted ${newCount} new articles`)
  return newCount
}

/* ── Stage 2: Cluster + synthesize (batch-limited for Vercel timeouts) ── */
export async function clusterAndSynthesize(maxClustersPerRun = 5): Promise<{ created: number; total: number }> {
  const cutoff = new Date(Date.now() - 48 * 3600 * 1000)

  // Fetch all unclustered recent articles
  const recent = await db
    .select({
      id:           rawArticles.id,
      title:        rawArticles.title,
      summary:      rawArticles.summary,
      url:          rawArticles.url,
      imageUrl:     rawArticles.imageUrl,
      sourceSlug:   sources.slug,
      sourceName:   sources.name,
    })
    .from(rawArticles)
    .innerJoin(sources, eq(sources.id, rawArticles.sourceId))
    .where(and(eq(rawArticles.clustered, false), gte(rawArticles.scrapedAt, cutoff)))
    .orderBy(desc(rawArticles.scrapedAt))

  console.log(`Unclustered articles (last 48h): ${recent.length}`)
  if (recent.length < 2) return { created: 0, total: 0 }

  const idToRow = new Map(recent.map(r => [r.id, r]))

  const inputs: ArticleInput[] = recent.map(r => ({
    id:         r.id,
    title:      r.title,
    summary:    r.summary ?? '',
    sourceSlug: r.sourceSlug,
  }))

  const clusters = runClustering(inputs)
  const toProcess = clusters.slice(0, maxClustersPerRun)
  let created = 0

  for (const cluster of toProcess) {
    const artsForSynth: ArticleForSynthesis[] = cluster.articleIds
      .map(id => idToRow.get(id))
      .filter((x): x is NonNullable<typeof x> => x !== undefined)
      .map(r => ({
        sourceName: r.sourceName,
        sourceSlug: r.sourceSlug,
        title:      r.title,
        summary:    r.summary ?? '',
        url:        r.url,
      }))

    const synth = await synthesizeCluster(artsForSynth)
    if (!synth) {
      console.warn(`Synthesis failed for cluster of ${artsForSynth.length} articles`)
      continue
    }

    // Find an image: from cluster articles, or fetch og:image from first one
    let imageUrl: string | null = null
    for (const id of cluster.articleIds) {
      const r = idToRow.get(id)
      if (r?.imageUrl) { imageUrl = r.imageUrl; break }
    }
    // og:image fallback skipped — too slow for Vercel 10s limit

    const [nc] = await db.insert(newsClusters).values({
      title:        synth.title.slice(0, 500),
      synthesis:    synth.synthesis,
      keyFacts:     synth.keyFacts,
      category:     synth.category,
      publishedAt:  new Date(),
      sourceCount:  new Set(cluster.sourceSlugs).size,
      imageUrl,
    }).returning({ id: newsClusters.id })

    const saBySlug = Object.fromEntries(synth.sourceAnalyses.map(sa => [sa.sourceSlug, sa]))

    for (const artId of cluster.articleIds) {
      const r = idToRow.get(artId)
      if (!r) continue
      const sa = saBySlug[r.sourceSlug]
      await db.insert(clusterArticles).values({
        clusterId:          nc.id,
        articleId:          artId,
        coveragePercentage: sa?.coveragePercentage ?? 0,
        emphasis:           sa?.emphasis ?? '',
        omissions:          sa?.omissions ?? '',
        similarityScore:    cluster.similarityScores[artId] ?? 0,
      }).onConflictDoNothing()
      await db.update(rawArticles).set({ clustered: true }).where(eq(rawArticles.id, artId))
    }

    created++
  }

  console.log(`Created ${created} clusters (of ${clusters.length} candidates)`)
  return { created, total: clusters.length }
}

/* ── Combined runner used by the cron endpoint ───────────────────── */
export async function runScrapingPipeline(): Promise<{ scraped: number; clustersCreated: number }> {
  const scraped = await ingestNewArticles()
  const { created } = await clusterAndSynthesize(1)
  return { scraped, clustersCreated: created }
}
